liquidctl/driver/kraken_three_x.py

```diff
@@ -0,0 +1,132 @@
+
+ liquidctl is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ liquidctl is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ """
+
+import logging
+
+from liquidctl.driver.usb import UsbHidDriver
+
+LOGGER = logging.getLogger(__name__)
+
+_READ_LENGTH = 64
+_WRITE_LENGTH = 64
+_MAX_READ_ATTEMPTS = 12
+
+
+class KrakenThreeX(UsbHidDriver):
+    """liquidctl driver for Kraken X3 devices from NZXT."""
+
+    SUPPORTED_DEVICES = [
+        (0x1e71, 0x2007, None, 'NZXT Kraken X3 Pump (X53, X63 or X73) (experimental)', {})
+    ]
+
+    def initialize(self, **kwargs):
+        """Initialize the device.
+
+        Reports the current firmware of the device.
+
+        Returns a list of (key, value, unit) tuples.
+        """
+        self.device.clear_enqueued_reports()
+        # initialize (don't work currently - responds with a 0x71 0x01 message?)
+        #self._write([0x70, 0x02, 0x01, 0xb8, 0x0b])
+        #self._write([0x70, 0x01])
+
+        # request static infos
+        self._write([0x10, 0x01])  # firmware info
+        status = []
+
+        def parse_firm_info(msg):
+            fw = '{}.{}.{}'.format(msg[0x11], msg[0x12], msg[0x13])
+            status.append(('Firmware version', fw, ''))
+
+        self._read_until({b'\x11\x01': parse_firm_info})
+        self.device.release()
+        return sorted(status)
+
+    def get_status(self, **kwargs):
+        """Get a status report.
+
+        Returns a list of `(property, value, unit)` tuples.
+        """
+        msg = self._read()
+
+        return [
+            ('Liquid temperature', msg[15] + msg[14] / 10, '°C'),
+            ('Pump speed', msg[18] << 8 | msg[17], 'rpm'),
+            ('Pump duty', msg[19], '%'),
+        ]
+
+    def set_fixed_speed(self, channel, duty, **kwargs):
+        """Set channel to a fixed speed duty."""
+        if channel != 'pump':
+            assert False, 'kraken X3 devices only support changing pump speeds'
+        if duty < 20 or duty > 100:
+            assert False, f'invalid duty value: {duty}. must be between 20 and 100!'
+
+        def parse_pump_speed(msg):
+            if msg[19] == duty:
+                LOGGER.debug(f'pump duty successfully changed to {msg[19]} % [{hex(msg[19])}]')
+            else:
+                assert False, f'pump duty did not update! currently at {msg[19]} % [{hex(msg[19])}]'
+
+        self._write([0x72, 0x01, 0x00, 0x00] + [duty] * 40)
+        self._read_until({b'\x75\x02': parse_pump_speed})
```

Comment:
I don't think this is necessary.

Reply:
Since the pump defaults to only 20% after a power cycle, I would like to know if for whatever reason the pump did not apply my changes. I hadn't a problem with this for now but better safe than sorry. 

What is the preferred way to throw an error? I've found many different implementations in other classes like this `assert` one, `raise ValueError()` or `LOGGER.error`.

Reply:
I understand your point, but this check doesn't give much assurance.  It's also a bit costly, particularly for API users or the new `liquidctl control` modes.¹

The [false positive reported in the issue thread](https://github.com/jonasmalacofilho/liquidctl/pull/100#issuecomment-619453674) tells me that when we read the duty the controller has simply copied back whatever duty value it stored in its memory before.  There's a chance an incoming HID report is enqueued just before the write, causing us to read stale data, resulting in the false positive.

But the duty being stored in the controller's memory is no real assurance or proper operation, particularly if we're talking about controller failures.  In most cases the controller will simply fail to ACK the write, and we'll get an `USBError` there.  A more obscure problem that could bypass that implicit check could just as well remain undetected by the subsequent read.

Again, I agree that blindly trusting the cooler carries some risk.  But dealing with that should be left to the user , and it's possible that many already have their own methods and limits for dealing with that.  Personally, I log a complete status 10 seconds after my boot-time configuration script has run (for debugging) and keep a display of the liquid temperature always visible (as that single piece of information can tell me a lot about the overall heath of my cooling system).

TL; DR.  This is more likely to generate false positives than actual positives.  Fixing this would make it even more costly, so I really think it should be removed.

¹ The work-in-progress modes I mentioned are related to #72 , #101 and the ['integrate-yoda'](https://github.com/jonasmalacofilho/liquidctl/tree/integrate-yoda) branch.

Reply:
Ok makes sense and I did remove it!